daemon/redis.c

```diff
@@ -1363,6 +1677,160 @@ static int redis_link_maps(struct redis *r, struct call *c, struct redis_list *m
 	return 0;
 }
 
+static void json_restore_call(struct redis *r, struct callmaster *m, redisReply *id, enum call_type type) {
+	redisReply* rr_jsonStr;
+	struct redis_hash call;
+	struct redis_list tags, sfds, streams, medias, maps;
+	struct call *c = NULL;
+	str s ;
+	const char *err = 0;
+	int i;
+#if !(GLIB_CHECK_VERSION(2,36,0))
+	g_type_init();
+#endif
+	JsonReader *root_reader =0;
+	JsonParser *parser =0;
+	char orig_json_callid[512]; memset(orig_json_callid,0,512);
+	memcpy(orig_json_callid,id->str,strlen(id->str));
+
+	rr_jsonStr = redis_get(r, REDIS_REPLY_STRING, "GET %s",id->str);
+	if (!rr_jsonStr) {
+		rlog(LOG_ERR, "Could not retrieve json data from redis for callid: %s", id->str);
+		goto err1;
+	}
+
+	// strip off json- prefix from callid
+	str_cut(id->str, 0, strlen("json-"));
+	id->len = strlen(id->str);
+	str_init_len(&s, id->str, id->len);
+
+	parser = json_parser_new();
+	if (!json_parser_load_from_data (parser, rr_jsonStr->str, -1, NULL)) {
+		rlog(LOG_DEBUG, "Could not parse json data !");
+		goto err1;
+	}
+	root_reader = json_reader_new (json_parser_get_root (parser));
+	if (!root_reader) {
+		rlog(LOG_DEBUG, "Could not read json data !");
+		goto err1;
+	}
+
+	c = call_get_or_create(&s, m, type);
+	err = "failed to create call struct";
+	if (!c)
+		goto err1;
+
+	c->root_reader = root_reader; // attach the json to the call in order to restore data from there
```

Comment:
I'm not sure I like putting the root_reader into the call object like this. I understand why you're doing it since it needs to be accessible, but it doesn't really seem to belong into the call struct. Maybe there's a better way to achieve the same thing.